### ISSUE TITLE
bitfinex live trading w/ simulated orders error - "bitfinex does not have market symbol <pair>"

### ISSUE THREAD

**Issue by gatapia** [enigmampc/catalyst#157]
My trading algo works well in non-live mode but changing to `live=True` and `simulate_orders=True` causes errors.  The `fetch_ticker` internal function throws the error: `bitfinex does not have market symbol DSH/USD`.  

Its also really hard to work in live mode as its hampered with these errors constantly:
`bitfinex GET https://api.bitfinex.com/v1/pubticker/<pair> 429 Too Many Requests  { "error": "ERR_RATE_LIMIT" }`.  Couldnt we be using the private API (with the users API/Secret) rather than the public API?  Or are tickers always public api?  

Note: I have > 10 coins in my algorithm so I assume its getting tickers for each and causing this.  

After dropping coins down to 5 I still get this error from the bitfinex API: `bitfinex POST https://api.bitfinex.com/v1/orders 429 Too Many Requests`.

I am using `catalyst, version 0.4.0` according to `catalyst --version`.

Note: All of this worked fine in `live=False` mode.

Cheers,
Guido

**Comment by fredfortier** [comment]
> The fetch_ticker internal function throws the error: bitfinex does not have market symbol DSH/USD
Does this error occur only with this particular ticker or all of the 10 markets in your algo.

> Note: I have > 10 coins in my algorithm so I assume its getting tickers for each and causing this.
Yes, we were initially using a `fetchTickers` method returning all tickers at once, but we temporarily switched over to a loop of `fetchTicker` after find some inconsistency in the data being returned. I these underlying issues should be fixed now, so we will revert back to `fetchTickers` which should solve your issue with the request number.

Consider upgrading to the latest release as we do frequent releases to quickly address reported issues.

**Comment by gatapia** [comment]
@fredfortier

I updated catalyst and now these are the errors I get:
```
[2018-01-11 23:13:26.115186] WARNING: CCXT: unable to fetch ticker bitfinex / iot_usd: bitfinex does not have market symbol IOT/USD
[2018-01-11 23:13:26.117325] WARNING: DataPortalExchange: fetching spot value again.
[2018-01-11 23:13:32.124879] WARNING: CCXT: unable to fetch ticker bitfinex / iot_usd: bitfinex does not have market symbol IOT/USD
[2018-01-11 23:13:32.125824] WARNING: DataPortalExchange: fetching spot value again.
[2018-01-11 23:13:40.627222] WARNING: CCXT: unable to fetch ticker bitfinex / iot_usd: bitfinex does not have market symbol IOT/USD
[2018-01-11 23:13:40.628536] WARNING: DataPortalExchange: fetching spot value again.
[2018-01-11 23:13:51.891122] WARNING: CCXT: unable to fetch ticker bitfinex / iot_usd: bitfinex does not have market symbol IOT/USD
[2018-01-11 23:13:51.892442] WARNING: DataPortalExchange: fetching spot value again.
[2018-01-11 23:14:07.775575] WARNING: CCXT: unable to fetch ticker bitfinex / iot_usd: bitfinex does not have market symbol IOT/USD
```

Followed by this exception:
```
/home/guido/anaconda3/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/exchange.pyc in get_spot_value(self, assets, field, dt, data_frequency)
    359             raise KeyError('Invalid column: {}'.format(field))
    360 
--> 361         tickers = self.tickers(assets)
    362         if field == 'close' or field == 'price':
    363             return [tickers[asset]['last'] for asset in tickers]

/home/guido/anaconda3/envs/catalyst/lib/python2.7/site-packages/catalyst/exchange/ccxt/ccxt_exchange.pyc in tickers(self, assets)
    864                 )
    865             )
--> 866             raise ExchangeRequestError(error=e)
    867 
    868         return tickers

ExchangeRequestError: Request failed: bitfinex does not have market symbol IOT/USD
```
Cheers
Guido

**Comment by fredfortier** [comment]
Ok thanks. I believe that it's a dup of issue #120 where CCXT is mapping USD => USDT but we are still mapping our symbols based on the exchange itself (Bitfinex uses USDT). We have this taken care of (simplifying the exchange config) in a separate branch which we will merge in the next few days. We'll update this issue as soon as it is done.

**Comment by kroitor** [comment]
Hi!

Here's what I know on this subject:
- DSH is DASH, but Bitfinex uses a non-standard code, despite that DASH is much more common than DSH.
- USD on Bitfinex stands for USDT, so appending that `T` reflects the reality better, especially in the light of the events they were going through recently (moved all their presence out of US, afaik).

Hope it helps. Thx!

**Comment by lacabra** [comment]
I also experience this same error with another related pair:
```
ExchangeRequestError: Request failed: bitfinex does not have market symbol IOT/ETH
```
So it may have to do more with `IOT`

**Comment by lacabra** [comment]
Of course, the short term fix is to leave out the troublesome coins. In case it helps, if you are generating the list of coins dynamically, and say you have a `coins` array of Catalyst symbol objects, you can easily filter out by doing:
```
coins = [c for c in coins if c.base_currency!='iot']
```

**Comment by lacabra** [comment]
I'm also getting `ratelimit` errors while trying to fetch `data.history()` for a set of coins:
```
ccxt.base.errors.DDoSProtection: bitfinex GET https://api.bitfinex.com/v2/candles/trade:30m:tSANETH/hist?sort=1&start=1515748800000&limit=21 429 Client Error: Too Many Requests for url: https://api.bitfinex.com/v2/candles/trade:30m:tSANETH/hist?sort=1&start=1515748800000&limit=21 ["error", 11010, "ratelimit: error"]
```
You may consider switching to another exchange. I've had better luck with others :)

**Comment by kroitor** [comment]
> So it may have to do more with IOT

The common symbol for that currency is IOTA, with the "A" ending, so ccxt does substitute for that, according to coinmarketcap and the majority of other exchanges (this and other substitutions done by Bitfinex,  like DSH → DASH, QTM → QTUM are here: https://github.com/ccxt/ccxt/blob/master/js/bitfinex.js#L218)

> I'm also getting ratelimit errors while trying to fetch data.history() for a set of coins

Bitfinex also has a very specific non-documented rate limit on the historical OHLCVs, but without the docs, we are still reverse-engineering the exact value.

My 2 cents...

**Comment by fredfortier** [comment]
The current issue is that we are in a transitory state where we are still creating the symbols from the exchanges without the CCXT substitution. We have a clean solution which does not involve mappings on a different branch but, in the interest of time, I will add some simple mapping to get past this issue.

**Comment by fredfortier** [comment]
> I'm also getting ratelimit errors while trying to fetch data.history() for a set of coins

This is another issue in the same category. We were initially fetching tickers with `fetch_tickers()` but temporarily switched to a loop of `fetch_ticker()` to work around some inconsistencies (which have been fixed since) between the two operations on Binance.

**Comment by kroitor** [comment]
@fredfortier on the fetch_tickers for Binance – we're still researching to workaround the issue. In the end we may return to a rate-limited, but a complete endpoint, I'll keep you guys updated, of course, will let you know beforehand. Thx!

**Comment by fredfortier** [comment]
@kroitor I was referring to this issue: https://github.com/ccxt/ccxt/pull/907, because of which we decided to temporarily loop through `fetch_ticker()` for each market. I think that this is fixed now so we can re-attempt to use the `fetch_tickers()` request. My understand that the rate-limit issue comes from the multiple request when requesting a bag of markets. Bitfinex seems to be particularly sensitive to this.

**Comment by fredfortier** [comment]
Fixed on 0.4.6 please re-open if you still notice an issue.
